orc/src/main/java/org/apache/iceberg/orc/ORCSchemaUtil.java

```diff
@@ -403,114 +398,150 @@ private static boolean isRequired(TypeDescription orcType) {
+      Types.StructType structType = Types.StructType.of(
+          fields.stream().filter(Optional::isPresent).map(Optional::get).collect(Collectors.toList()));
+      return Optional.of(getIcebergType(icebergIdOpt.get(), icebergToOrcMapping.get(icebergIdOpt.get()).name(),
+          structType, isRequired));
+    }
 
-        return getIcebergType(icebergID, name, mapTypeWithKV, isRequired);
+    @Override
+    public Optional<Types.NestedField> list(Types.ListType iList, TypeDescription array,
+                                            Optional<Types.NestedField> element) {
+      boolean isRequired = isRequired(array);
+      Optional<Integer> icebergIdOpt = icebergID(array);
+
+      if (!icebergIdOpt.isPresent() || !element.isPresent()) {
+        return Optional.empty();
       }
-      default:
-        // We don't have an answer for union types.
-        throw new IllegalArgumentException("Can't handle " + orcType);
+
+      Types.NestedField foundElement = element.get();
+      Types.ListType listTypeWithElem = isRequired(array.getChildren().get(0)) ?
+          Types.ListType.ofRequired(foundElement.fieldId(), foundElement.type()) :
+          Types.ListType.ofOptional(foundElement.fieldId(), foundElement.type());
+      return Optional.of(getIcebergType(icebergIdOpt.get(),
+          icebergToOrcMapping.get(icebergIdOpt.get()).name(), listTypeWithElem, isRequired));
     }
-  }
 
-  private static int getMaxIcebergId(TypeDescription originalOrcSchema) {
-    int maxId = icebergID(originalOrcSchema).orElse(0);
-    final List<TypeDescription> children = Optional.ofNullable(originalOrcSchema.getChildren())
-        .orElse(Collections.emptyList());
-    for (TypeDescription child : children) {
-      maxId = Math.max(maxId, getMaxIcebergId(child));
+    @Override
+    public Optional<Types.NestedField> map(Types.MapType iMap, TypeDescription map, Optional<Types.NestedField> key,
+                                           Optional<Types.NestedField> value) {
+      boolean isRequired = isRequired(map);
+      Optional<Integer> icebergIdOpt = icebergID(map);
+
+      if (!icebergIdOpt.isPresent() || !key.isPresent() || !value.isPresent()) {
+        return Optional.empty();
+      }
+
+      Types.NestedField foundKey = key.get();
+      Types.NestedField foundValue = value.get();
+      Types.MapType mapTypeWithKV = isRequired(map.getChildren().get(1)) ?
+          Types.MapType.ofRequired(foundKey.fieldId(), foundValue.fieldId(), foundKey.type(), foundValue.type()) :
+          Types.MapType.ofOptional(foundKey.fieldId(), foundValue.fieldId(), foundKey.type(), foundValue.type());
+
+      return Optional.of(getIcebergType(icebergIdOpt.get(), icebergToOrcMapping.get(icebergIdOpt.get()).name(),
+          mapTypeWithKV, isRequired));
     }
 
-    return maxId;
+    @Override
+    public Optional<Types.NestedField> primitive(Type.PrimitiveType iPrimitive, TypeDescription primitive) {
```

Comment:
Would it make sense to define a `OrcSchemaVisitor` visitor, similar to `org.apache.iceberg.avro.AvroSchemaVisitor`. This is would the Iceberg type parameter passed in. Here we are not making use of the Iceberg type parameter at all so I wonder. if that's a little confusing.

Reply:
Yeah, I think I can add a `OrcSchemaVisitor` for this case.